poly1305/src/backend/avx2.rs

```diff
@@ -60,10 +68,13 @@ impl State {
         }
     }
 
-    /// Reset internal state
-    pub(crate) fn reset(&mut self) {
-        self.initialized = None;
-        self.num_cached_blocks = 0;
+    /// Process four Poly1305 blocks at once.
+    #[target_feature(enable = "avx2")]
+    pub(crate) unsafe fn compute_par_blocks(&mut self, blocks: &ParBlocks) {
+        assert!(self.partial_block.is_none());
+        assert_eq!(self.num_cached_blocks, 0);
```

Comment:
Hmm, now that I've fixed the bytes-vs-blocks confusion and the parallel blocks code is actually triggering, AEAD tests fail with this branch because the AD is added to the MAC before the ciphertext, so unless the AD length is `[(4k + 3) * BLOCK_SIZE..(4k + 4) * BLOCK_SIZE]`, the parallel blocks won't line up.

We can handle this case specifically in the AEAD, but we should think about a way to handle this better in the traits. Currently users will just hit panics, but if we handled this off-by-one by falling back to non-parallelized, then it becomes very easy to not trigger the parallelism by accident.

Reply:
Added a suggested change to `UhfBackend`: https://github.com/RustCrypto/traits/pull/965#discussion_r857126434